test/pages/register-a-beacon/check-beacon-details.test.tsx

```diff
@@ -32,3 +34,133 @@ describe("CheckBeaconDetails", () => {
+      manufacturer: "Test manufacturer",
+      model: "Test model",
+      hexId: "",
+    };
+    const expectedErrorMessages = [expect.stringContaining("required")];
+    const formManager = definePageForm(fieldValues);
+    formManager.markAsDirty();
+
+    const validationResult = formManager.serialise();
+
+    expect(validationResult.hasErrors).toBe(true);
+    expect(validationResult.errorSummary.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages).toEqual(
+      expect.arrayContaining(expectedErrorMessages)
+    );
+  });
+
+  it("should error if hex string is not 15 characters long", () => {
+    const fieldValues = {
+      manufacturer: "Test manufacturer",
+      model: "Test model",
+      hexId: ukHexIdWith1ExtraCharacter,
+    };
+    const expectedErrorMessages = [expect.stringContaining("15 characters")];
+    const formManager = definePageForm(fieldValues);
+    formManager.markAsDirty();
+
+    const validationResult = formManager.serialise();
+
+    expect(validationResult.hasErrors).toBe(true);
+    expect(validationResult.errorSummary.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages).toEqual(
+      expect.arrayContaining(expectedErrorMessages)
+    );
+  });
+
+  it("should error if hexId string is not hexadecimal", () => {
+    const fieldValues = {
+      manufacturer: "Test manufacturer",
+      model: "Test model",
+      hexId: stringWithNonHexadecimalChars,
+    };
+    const expectedErrorMessages = [
+      expect.stringContaining("numbers 0 to 9 and letters A to F"),
+      expect.stringContaining("UK-encoded"),
+    ];
+    const formManager = definePageForm(fieldValues);
+    formManager.markAsDirty();
+
+    const validationResult = formManager.serialise();
+
+    expect(validationResult.hasErrors).toBe(true);
+    expect(validationResult.errorSummary.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages.length).toBe(2);
+    expect(validationResult.fields.hexId.errorMessages).toEqual(
+      expect.arrayContaining(expectedErrorMessages)
+    );
+  });
+
+  it("should error if hexId is not UK-encoded", () => {
+    const fieldValues = {
+      manufacturer: "Test manufacturer",
+      model: "Test model",
+      hexId: validOtherCountryHexId,
+    };
+    const expectedErrorMessages = [expect.stringContaining("UK-encoded")];
+    const formManager = definePageForm(fieldValues);
+    formManager.markAsDirty();
+
+    const validationResult = formManager.serialise();
+
+    expect(validationResult.hasErrors).toBe(true);
+    expect(validationResult.errorSummary.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages.length).toBe(1);
+    expect(validationResult.fields.hexId.errorMessages).toEqual(
+      expect.arrayContaining(expectedErrorMessages)
+    );
+  });
+});
```

Comment:
Contrast this test of the hexId validation _logic_ with the UI test of how the errors are _displayed_ in [checkBeaconDetails.spec.ts](https://github.com/mcagov/beacons-webapp/blob/f449b7489ccf3ecdb11c6f2c46c4fee2affad03d/cypress/integration/checkBeaconDetailsPage.spec.ts).

These tests _do_ check for strings in error messages, but the concern that is being tested is "what type of error message is this" not "what does the error message say".  (Hence the strings being minimal: "required" and "UK-encoded")

These tests also comprehensively checks the logic of the hexId field, covering all the reasons/cases it might be invalid.  It is very quick to run, so the quick developer feedback loop when developing complex/interdependent form logic is supported.

It also makes sure the `validators` are not stepping on top of each other.  For example, if a field has both `required` and `isLength` validators applied, this test makes sure only one of them fires for blank inputs.  This is not covered in `FormManager`'s unit tests, where mocking the validators is necessary.